src/green/gpu/cugw_qpt.h

```diff
@@ -265,30 +265,74 @@ namespace green::gpu {
      * \param Pqk_tQP Dressed polarization bubble
      */
-    void compute_second_tau_contraction(cxx_complex* Sigmak_stij_host = nullptr, cuda_complex* Pqk_tQP = nullptr);
+    void compute_second_tau_contraction(cuda_complex* Pqk_tQP = nullptr);
+
     /**
      * \brief Using dressed GW polarization compute self-energy at a given momentum point (X2C version)
-     * \param Sigmak_stij_host Host stored array for Self-energy at a given momentum point
+     * 
      * \param Pqk_tQP Dressed polarization bubble
      */
-    void compute_second_tau_contraction_2C(cxx_complex* Sigmak_stij_host = nullptr, cuda_complex* Pqk_tQP = nullptr);
+    void compute_second_tau_contraction_2C(cuda_complex* Pqk_tQP = nullptr);
 
     /**
      * \brief For a given k-point copy self-energy back to a host memory
      * \param low_memory_mode - whether the whole self-energy allocated in memory or not
+     */
+    void write_sigma(bool low_memory_mode = false);
+
+    /**
+     * \brief return the status of copy_selfenergy from device to host
+     * 
+     * \return false - not required, stream ready for next calculation
+     * \return true - required
+     */
+    bool require_cleanup(){
+      return cleanup_req_; 
+    }
+
+    /**
+     * \brief perform cleanup, i.e. copy data from Sigmak buffer (4-index array for a given momentum point) to Host shared memory Self-energy
+     * 
+     * \param low_memory_mode - whether the whole self-energy allocated in memory or not
      * \param Sigmak_stij_host - Host stored self-energy object at a given momentum point
+     * \param Sigma_tskij_host - Host stored full self-energy object for each device
+     * \param x2c - use x2c specific functions or not
+     */
+    void cleanup(bool low_memory_mode, tensor<std::complex<prec>, 4>& Sigmak_stij_host, ztensor<5>& Sigma_tskij_host, bool x2c);
+
+    /**
+     * \brief Copy the non-relativistic self-energy from the per-k buffer to the full host tensor.
+     * 
+     * \param Sigma_tskij_host Host storage for the full self-energy tensor.
+     * \param Sigmak_stij Self-energy tensor for a given momentum point.
+     */
+    void copy_Sigma(ztensor<5>& Sigma_tskij_host, tensor<std::complex<prec>, 4>& Sigmak_stij);
+
+    /**
+     * \brief Copy 2-component self-energy from per-k buffer to full host tensor.
+     * 
+     * \param Sigma_tskij_host Host storage for the full self-energy tensor.
+     * \param Sigmak_stij Self-energy tensor for a given momentum point.
      */
-    void write_sigma(bool low_memory_mode = false, cxx_complex* Sigmak_stij_host = nullptr);
+    void copy_Sigma_2c(ztensor<5>& Sigma_tskij_host, tensor<std::complex<prec>, 4>& Sigmak_stij);
 
     /**
-     * \brief Check if cuda devices are budy
+     * \brief Check if cuda devices are busy
      * \return true if asynchronous calculations are still running
      */
     bool is_busy();
 
+    /**
+     * \brief Set the k_red_id_ object
```

Comment:
propose a better explanation of what 'k_red_id' stands for, since a new reader won't know.